Lib/concurrent/futures/_base.py

```diff
@@ -386,7 +386,11 @@ def done(self):
 
     def __get_result(self):
         if self._exception:
-            raise self._exception
+            try:
+                raise self._exception
+            finally:
+                # Break a reference cycle with the exception in self._exception
+                self = None
```

Comment:
Did you mean "self._exception = None"?

Reply:
No, because the exception **must be** available to the caller via "future.result()" or "future.exception()". We can not drop the exception reference.

When the caller calls "future.result()" and a exception was recorded, the exception is reraised in the new context. That context encapsulates the old exception raised in the "future" context but also current context, including "self". Since "self" has a "self._exception" field with the "future" exception, we have a reference loop.

We can not delete "self._exception" because the exception need to be reported and could be accessed multiple times by the caller (you can call "future.result()" multiple times. So I do "self = None" to break the loop. This have no bad side effects because "self" goes out of scope ant the end of the call, but it would be be keep alive because the cycle I am breaking... for that very reason.

Please, check the original bug report.

I have some block diagrams:

Before:

https://lists.es.python.org/pipermail/general/attachments/20201229/0c14bc58/attachment-0002.png

Notice the loop in the left bottom block (_base.py:442) to the future. That prevents timely collection. The loop from the dict in the right bottom corner is an artifact of the data collection.

After:
https://lists.es.python.org/pipermail/general/attachments/20201229/0c14bc58/attachment-0003.png

Here we don have a loop anymore . The loop from dict is an artifact of the data collection.

Reply:
In https://lists.es.python.org/pipermail/general/attachments/20201229/0c14bc58/attachment-0002.png , loops from "_base.py:389" and "_base.py:442" are what I am braking with "self = None".

Reply:
I will mark the conversation as "resolved" now. Feel free to reopen if you think that your point has not been addressed appropriately.

Thanks for your evaluation.

Reply:
That's fine, I understand now. Thank you.